### DISCUSSION TITLE
Help with keyset pagination for projects

### DISCUSSION THREAD

**Discussion by m-rusu** [ubisoft/NGitLab#808] [Q&A] [answered]
Can anyone help in explaining how keyset pagination works for the GET Projects endpoint?
I am using it like this:

`var query = new ProjectQuery()`
`{`
    `OrderBy = "id",`
    `PerPage = pageSize,`
    `Ascending = true`
`};`

`var enumerator = client.Projects.GetAsync(query).GetAsyncEnumerator(cancellationToken);`

And then I'm going through the enumerator and adding the current project to a list, until the count is higher than the pageSize. How do I go about retrieving the next page? 

In the  `GetAsyncEnumerator`, only the items are returned, not the links to the next, previous, first or last pages.

**Comment by louis-z** [comment, marked as answer]
Hi @m-rusu,

The call to `IProjectClient.GetAsync(ProjectQuery query)` returns a `GitLabCollectionResponse<Project>`, which can be directly enumerated synchronously or asynchronously. Handling of GitLab pages is done under the hood, so you don't need to worry about accessing successive pages yourself. Specifying `ProjectQuery.PerPage` is not required (it will default to 20), but you may get better results in certain cases if you do.

For instance, say you wanted to get the first 83 projects. The optimal way would probably be to
- tell GitLab you want 83-project pages, and
- limit enumeration over the first 83 projects (i.e. the first page returned by GitLab):

```cs
var perPage = 83;
var query = new ProjectQuery()
{
    OrderBy = "id",
    PerPage = perPage,
    Ascending = true
};

foreach (var project in client.Projects.GetAsync(query).Take(perPage))
{
}
```

Note that the call to `Take()` is necessary to limit enumeration to the selected number of projects. Without that call, you would end up enumerating all projects.

**Comment by m-rusu** [comment]
Alright, thank you for the response!
